### PR TITLE
adding dynamic market system complete with config options

### PR DESCRIPTION
**Pull request by dbenson24**
reference #47 for discussion

squashed PR

### CODE REVIEW

**Comment by DarkSilencerZA** [comment]
@dbenson24 Found an edited / fixed version of RyanTTs.

https://native-network.net/thread/2960-tutorial-altis-life-4-4r3-ryantts-dynamic-market-system/

Any help for you?

**Comment by dbenson24** [comment]
@DarkSilencerZA You no longer need the table at all for non persistent mode

**Comment by DarkSilencerZA** [comment]
Epic! Good one bro!

**Comment by dbenson24** [comment]
@BoGuu @tkcjesse @Jawshy is there anything that you want before this gets merged?

**Comment by tkcjesse** [comment]
Yeah, we need to adjust how the files are being modified. Rather than modify the existing we need to use exitWiths to essentially use replacement files.

**Comment by Jawshy** [comment]
> ... @Jawshy is there anything that you want before this gets merged?

Yes, updated BattlEye filters which I can do. Shall I push to [your fork](https://github.com/dbenson24/Framework-1/tree/dynamicmarket) in a PR?

File: life_server/Functions/Economy/fn_loadPrices.sqf [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                _sellPrice,
+                ITEM_VARPRICE(_varName),
+                ITEM_MINPRICE(_varName),
+                ITEM_MAXPRICE(_varName),
+                ITEM_FACTOR(_varName),
+                _sellPrice,
+                _sellPrice
+            ];
+    missionNamespace setVariable [_name, _info];
+    publicVariable _name;
```
Comment (NiiRoZz) [inline life_server/Functions/Economy/fn_loadPrices.sqf]: Don't use publicVariable, now you can set missionNamespace setVariable [_name, _info, true]; for auto publicvariable

File: life_server/Functions/Economy/fn_loadPrices.sqf [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+} forEach _queryResult;
+
+missionNamespace setVariable ["MarketPrices", _marketPrices];
+publicVariable "MarketPrices";
```
Comment (NiiRoZz) [inline life_server/Functions/Economy/fn_loadPrices.sqf]: Like above

**Comment by dbenson24** [comment]
@NiiRoZz is there any performance benefit to doing that? I feel like explicitly using the publicVariable command is clearer.

@Jawshy yea just make a PR to Framework-1

**Comment by NiiRoZz** [comment]
@dbenson24 Just more readable With my suggestion, but do what you want ..

**Comment by tkcjesse** [comment]
@dbenson24 he has this idea that it's more "optimized"... @BoGuu would remember this situation lol...

https://i.gyazo.com/d7b987120c8e67343bc26e25472557b6.png

**Comment by NiiRoZz** [comment]
@tkcjesse , In my language, "optimized" have 2 meaning,
- FIRST : Optimisation of any script with new command or rethink script.
- SECOND : Optimisation of reading any script.

All the time i have anyone speak me, "why you don't put comment for beginner scripter ?",
Yeah i see you never have this question, because all the time, you would want the person who is right, and you don't help any beginner, me i taje all the time to improve my knowledge and knowledge of anyone.

**Comment by Jawshy** [comment]
Can you [localise](https://community.bistudio.com/wiki/localize) [diag_log](https://community.bistudio.com/wiki/diag_log)?

**Comment by setoy** [comment]
I've applied this system to our server. However apples don't show up in our market whereas tbacon does.

Values in DB:
tbacon
> 39	tbacon	8	2	1	4	4	0

apple
> 40 apple	2	1	1	1	1	0

Is there some limitation regarding any of those fields that make the resource not display in the shops?

**Comment by dbenson24** [comment]
@setoy can you provide some more details?
Are you saying that apples can no longer be purchased at a market shop?

**Comment by setoy** [comment]
@dbenson24 
They can be purchased but not sold.

**Comment by dbenson24** [comment]
@setoy if the sell price is <1 it cannot be sold on a store

**Comment by setoy** [comment]
@dbenson24 
We've got it set to 1, yet it still didn't show up when selling. We had to set it to "2". Maybe theres some <= 1 somewhere? :p

**Comment by dbenson24** [comment]
@setoy okay I figured it out, I'll fix it later tonight

**Comment by DarkSilencerZA** [comment]
3:22:16 Error in expression <
diag_log format ["virtual items : %1", _items];

if ((getNumber(missionConfigFi>
 3:22:16   Error position: <_items];

if ((getNumber(missionConfigFi>
 3:22:16   Error Undefined variable in expression: _items
 3:22:16 File life_server\Functions\Economy\fn_syncPrices.sqf, line 10

**Comment by dbenson24** [comment]
@DarkSilencerZA @setoy fixed

**Comment by Jawshy** [comment]
No errors found as of https://github.com/AsYetUntitled/Framework/pull/64/commits/9324102cff39899a4619d017bad6d7278b2ad463.

**Comment by DarkSilencerZA** [comment]
I would suggest a check for the market.. Don;t want cops meta gaming etc.

if (playerSide isEqualTo civilian)then {};

**Comment by Jawshy** [comment]
@DarkSilencerZA, do you even [markdown](https://help.github.com/articles/basic-writing-and-formatting-syntax/#quoting-code)?

**Comment by dbenson24** [comment]
@DarkSilencerZA I might add that to the config options. I think that if civilians can so easily know the prices of everything, cops should too. But I know some people will disagree

**Comment by DarkSilencerZA** [comment]
@dbenson24 It's more of a meta-gaming thing. Cops can check the market and camp an active drug trader etc.

Otherwise we can just make a true / false in the config_master for it and people can decide what they want.

**Comment by tkcjesse** [comment]
Or you can just not be lazy and just add to the keyhandler if (playerside isEqto Civ) and be on your way. Like why does everything need a friggin option. What happened to people working on their own shit once and while?

**Comment by Jawshy** [comment]
I did another quick test and found no obvious errors. (https://github.com/AsYetUntitled/Framework/pull/64/commits/d21cc75528f4b5a286ed1d7730ffb8056d4471c5)

**Comment by sargken** [comment]
@dbenson24 Status?

**Comment by adonis17** [comment]
17:55:46 Error in expression <onConfigFile >> "VirtualItems" >> _good select 0 >> "displayName");

ctrlSetText>
17:55:46   Error position: <select 0 >> "displayName");

ctrlSetText>
17:55:46   Error Erreur générique dans une expression
17:55:46 File mpmissions\__CUR_MP.Tanoa\core\economy\fn_marketGrab.sqf, line 29
17:55:46 Error in expression <onConfigFile >> "VirtualItems" >> _good select 0 >> "displayName");

ctrlSetText>
17:55:46   Error position: <select 0 >> "displayName");

ctrlSetText>
17:55:46   Error Erreur générique dans une expression
17:55:46 File mpmissions\__CUR_MP.Tanoa\core\economy\fn_marketGrab.sqf, line 29

**Comment by dbenson24** [comment]
@adonis17 ce config paramètres utilisez-vous?

**Comment by adonis17** [comment]
@dbenson24 
`dynamic_market = true; //True enables Dynamic Market
    dynamic_market_persistence = true; //True makes price changes last across restarts
    end_time = 21600; //Last time to sync prices (Default: 4 hours)
    sync_rate = 900; //Seconds between syncs to database (Default: 15 minutes)
    adjustment_delay = 60; //Seconds it takes for price changes to hit the market (Default: 2 minutes)`

**Comment by setoy** [comment]
Will this eventualyl get merged?

@dbenson24

**Comment by dbenson24** [comment]
@setoy you're asking the wrong person

**Comment by Jawshy** [comment]
@setoy, no, because there are conflict that must be resolved among other things.

**Comment by setoy** [comment]
@dbenson24 Alright, thought this wasn't finished yet. Any thoughts on my suggestion to explain a few settings inside the new config options for vItems?

**Comment by dbenson24** [comment]
@setoy yea the settings should definitely be explained

**Comment by DarkSilencerZA** [comment]
This dyn market is a little broken... items hit rock bottom prices way too fast - even with the varPrice set to 1.

**Comment by Jason2605** [comment]
Edit the var price so they adjust less then i guess?

**Comment by dbenson24** [comment]
@DarkSilencerZA in line 66 of adjustprices.sqf you can change the coefficient that affects how much price changes. Right now it randomly bounces between 0.3 and 1.7, if you change this to 0.03 and .17 you should see changes in the market reduced by a factor of 10 across the board without playing with any of the varparices

**Comment by DarkSilencerZA** [comment]
@dbenson24 Thanks! Will give it a go.

**Comment by NeFShARk** [comment]
Nice, i thought that this was dead hehe >P

It's nice to see this coming back to life again! This is an important enhancement in my opinion :)

**Comment by dbenson24** [comment]
@Jawshy I checked it for you
@NeFShARk it's not that this was dead, it just seems that the community is split. Many people feel that building a 'framework' with tons of features will just lead to an even greater surplus of shitty server owners just trying to make a quick buck. Others feel that the framework should be built out a bit and have some more features by default. This branch is feature complete, and it's pretty well bug tested at this point. The only changes that one might wish to make are for code modularity and making it easier for someone with other market changes to interface with the way that I rewrote how virtual shops populate their menus. The PR is here, devs that want to be able to integrate a dynamic market into their server can do so with a quick marge.

**Comment by Jawshy** [comment]
I've made the changes, but GitHub won't show them.

File: life_server/config.cpp [hunk 3]
```diff
@@ -92,7 +92,14 @@ class CfgFunctions {
...
             class houseGarage {};
         };
-
+    
+        class Economy {
+            file = "\life_server\Functions\Economy";
+            class adjustPrices {};
+            class loadPrices {};
+            class syncPrices {};
+        }
```
Comment (NeFShARk) [inline life_server/config.cpp]: It's missing a ;

**Comment by DarkSilencerZA** [comment]
@dbenson24 whats the need for `_sellingfactor = (count _itemArray) - 1;`

**Comment by MypowerHD** [comment]
Hey Ho,

I have some issues with the Dynamic Market System. There is also an Closed issue written by me #201 for more info's!

> When i buy an item, the item costs 0.
When i disable the Dynamicmarket via "dynamic_market = false; //True enables Dynamic Market", all works right and the items have their current purchase price!

**Comment by dbenson24** [comment]
@DarkSilencerZA _sellingfactor increasing the price drops when there are a lot of items in the market, I did this to decrease inflation. Without it over time the market is going to have a much higher average price on everything.

@MypowerHD how did you integrate the dynamic market? did you use Git or do it by hand?

**Comment by MypowerHD** [comment]
@dbenson24, i coppy all required files and tooked into my files.

My struckture is:
. #Mission.Altis
├── CryZ
│   ├── core
│   └── dialog
├── CryZ_code #Folder 4 new scripts
│   └── economy                # Unit tests
└── ...

my Functions.hpp:
`class CryZ_code
{
    tag = "CryZ";
    class Economy {
        file="CryZ_code\economy";
        class marketGrab {};
        class openEconomy {};
        class retrievePrices {};
        class updateEconomy {};
        class updatePrice {};
    };

    class Functions
    {
        file = "CryZ_code\fnc";
        class teleportEx {};
    };
};`

update all function that called client side to the tag CryZ

**Comment by MypowerHD** [comment]
@dbenson24, you have no idea? :(

**Comment by coresync2k** [comment]
it is normal? i add t-item -> "test1" i add this into "market -> Config_Vitems" to buy, but there is nothing.
if i add this item to the "Database" it works? but without Database it works not?

**Comment by setoy** [comment]
@dbenson24 This is broken with extDB3. Do you have a working sample?

**Comment by coresync2k** [comment]
this is not broken with extdb3 O_o

**Comment by Larry-Lizard** [comment]
@dbenson24 Would be nice if you could fix the merge conflict so we could actually see this in game :)

**Comment by dbenson24** [comment]
@Larry-Lizard I believe the decision was made to not merge this and leave it as a PR for those who are interested because dynamic markets do not belong in the core game

**Comment by Larry-Lizard** [comment]
@dbenson24 Well it still has the tag, but if it works then good. Any docs on the implementation of that?

**Comment by Zevba** [comment]
12:07:02 Error in expression <ariable (_varName);
_itemArray pushBack _priceArray;
} forEach _shopItems;
} els>
12:07:02   Error position: <_priceArray;
} forEach _shopItems;
} els>
12:07:02   Error Undefined variable in expression: _pricearray
12:07:02 File core\economy\fn_retrievePrices.sqf [life_fnc_retrievePrices], line 76
12:07:02 Error in expression <ariable (_varName);
_itemArray pushBack _priceArray;
} forEach _shopItems;
} els>
12:07:02   Error position: <_priceArray;
} forEach _shopItems;
} els>
12:07:02   Error Undefined variable in expression: _pricearray
12:07:02 File core\economy\fn_retrievePrices.sqf [life_fnc_retrievePrices], line 76
12:07:07 Error in expression <ariable (_varName);

**Pull request closed without merging**